non-critical-infra/modules/vaultwarden.nix

```diff
@@ -0,0 +1,114 @@
+{ config, ... }:
+{
+  services.vaultwarden = {
+    enable = true;
+    backupDir = "/var/lib/bitwarden_rs/backups/";
+    environmentFile = "/var/lib/bitwarden_rs/vaultwarden.env";
+    config = {
+      DOMAIN = "https://vault.nixos.org";
+      SIGNUPS_ALLOWED = false;
+      SHOW_PASSWORD_HINT = false;
+      ROCKET_ADDRESS = "127.0.0.1";
+      ROCKET_PORT = 8222;
+      ROCKET_LOG = "critical";
+      SMTP_HOST = "mail.gandi.net";
+      SMTP_PORT = 587;
+      SMTP_SSL = true;
+      SMTP_FROM = "vaultwarden@nixos.org";
+      SMTP_FROM_NAME = "NixOS Vaultwarden";
+    };
+  };
+
+  services.nginx = {
+    enable = true;
+    recommendedGzipSettings = true;
+    recommendedOptimisation = true;
+    recommendedProxySettings = true;
+    recommendedTlsSettings = true;
+    virtualHosts."vault.nixos.org" = {
+      forceSSL = true;
+      enableACME = true;
+      locations."/" = {
+        proxyPass = "http://127.0.0.1:${toString config.services.vaultwarden.config.ROCKET_PORT}";
+      };
+    };
+  };
+
+  sops.secrets = {
+    vaultwarden-env = {
+      sopsFile = ../secrets/vaultwarden-env.caliban;
+      format = "binary";
+      path = "/var/lib/bitwarden_rs/vaultwarden.env";
+    };
+
+    vaultwarden-backup-pass = {
+      sopsFile = ../secrets/vaultwarden-backup-pass.caliban;
+      format = "binary";
+    };
+  };
+
+  services.borgbackup.jobs.vaultwarden = {
+    encryption = {
+      passCommand = "cat /run/secrets/vaultwarden-backup-pass";
+      mode = "repokey-blake2";
+    };
+    repo = "/mnt/backup-share/caliban.nixos.org/vaultwarden";
+    paths = [ "/var/lib/bitwarden_rs/backups/" ];
+    prune.keep = {
+      daily = 7;
+      weekly = 4;
+      monthly = -1; # Keep at least one archive for each month
+    };
+  };
+
+  services.fail2ban = {
+    enable = true;
+    jails = {
+      "vaultwarden-admin" = ''
+        enabled = true
+        port = 80,443
+        filter = vaultwarden-admin[journalmatch='_SYSTEMD_UNIT=vaultwarden.service']
+        maxretry = 3
+        bantime = 14400
+        findtime = 14400
+        backend = systemd
+      '';
+      "vaultwarden-web" = ''
+        enabled = true
+        port = 80,443,8081
+        filter = vaultwarden-web[journalmatch='_SYSTEMD_UNIT=vaultwarden.service']
+        maxretry = 3
+        bantime = 14400
+        findtime = 14400
+      '';
+    };
+  };
+
+
+  environment.etc = {
+    "fail2ban/filter.d/vaultwarden-admin.conf".text = ''
+      [INCLUDES]
+      before = common.conf
+
+      [Definition]
+      failregex = ^.*Invalid admin token\. IP: <ADDR>.*$
+      ignoreregex =
+
+    '';
+
+    "fail2ban/filter.d/vaultwarden-web.conf".text = ''
+      [INCLUDES]
+      before = common.conf
+
+      [Definition]
+      failregex = ^.*Username or password is incorrect\. Try again\. IP: <ADDR>\. Username:.*$
+      ignoreregex =
+    '';
+
+  };
```

Comment:
It is a bit hidden but you can write the filters into nixos options, too.

```
services.fail2ban.jails = {
          # https://github.com/dani-garcia/vaultwarden/wiki/Fail2Ban-Setup#setup-for-web-vault
          vaultwarden = {
            filter = {
              INCLUDES.before = "common.conf";
              Definition = {
                failregex = "^.*Username or password is incorrect\. Try again\. IP: <ADDR>\. Username:.*$";
                ignoreregex = "";
              };
            };
            settings = {
              backend = "systemd";
              port = "80,443";
              filter = "vaultwarden[journalmatch='_SYSTEMD_UNIT=vaultwarden.service']";
              banaction = "%(banaction_allports)s";
              maxretry = 3;
              bantime = 14400;
              findtime = 14400;
            };
          }
```

Reply:
Oh thanks, I also though that the way that I did it (and that is documented at multiple places) was very ugly. Will fix.

Reply:
Looks like that does not work:
```
services.fail2ban.jails.vaultwarden-admin' is not of type `strings concatenated with "\n"'
```

Reply:
Ah, I think this is only on unstable. I'll fix it after the release.

Reply:
Done that.